pkg/common-controller/snapshot_controller.go

```diff
@@ -778,8 +778,9 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotErrorStatusWithEvent(snap
 		Message: &message,
 	}
 	snapshotClone.Status.Error = statusError
-	ready := false
-	snapshotClone.Status.ReadyToUse = &ready
+	// NOTE(xyang): ReadyToUse in VolumeSnapshot's Status should only be updated based on
+	// ReadyToUse in VolumeSnapshotContent's Status. When ReadyToUse is already True
+	// in VolumeSnapshot's Status, we should not set it to false in updateSnapshotStatusWithEvent.
 	newSnapshot, err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshots(snapshotClone.Namespace).UpdateStatus(context.TODO(), snapshotClone, metav1.UpdateOptions{})
```

Comment:
@xing-yang  and I have discussed, and initially I agreed on this approach. In general I still think this is the right way to do things except one case which I am not 100% clear after looking into the code. It does have make the bi-directional binding verification more critical though. With this, situations like previously ready VolumeSnapshot resources will maintain the ready state forever(this seems to go against the "status represent the current state of a resource" principle)! Even in cases like losing binding from VolumeSnapshotContent, or VolumeSnapshotContent has been modified to refer to another VolumeSnapshot. In those cases, there is an error on this VolumeSnapshot and a short-lived event. I am a little concerned it might break existing user's expectation where they could rely on checking the ready flag (though they should not). Other than that, the PR LGTM.
@msau42 @saad-ali do you have any concerns?

Reply:
We could also add an input parameter "readyToUse" in updateSnapshotErrorStatusWithEvent() and let the caller decide whether to change it or not.

Reply:
To summarize, I think the safer approach would be something like:

- Do not set any status fields on VolumeSnapshot (except error) before binding is complete. Only after binding is complete we sync the other fields.
- Error/event setting methods only set error and events, don't modify any other fields

Reply:
> To summarize, I think the safer approach would be something like:
> 
> * Do not set any status fields on VolumeSnapshot (except error) before binding is complete. Only after binding is complete we sync the other fields.

>> This is already the logic, problem is once the ready == true, the controller does binding checks and based on the situation using this exact call to update status in the cases where binding is invalid (either content has been deleted or content has been modified to point to another snapshot) [code](https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/common-controller/snapshot_controller.go#L399). With this change, the ready flag will *not* be modified even though for those case.

> * Error/event setting methods only set error and events, don't modify any other fields

>> agreed

>

I am good with the change in general, at least we need to document the behavior change in the original issue.

Reply:
> This is already the logic, problem is once the ready == true, the controller does binding checks and based on the situation using this exact call to update status in the cases where binding is invalid (either content has been deleted or content has been modified to point to another snapshot) code. With this change, the ready flag will not be modified even though for those case.

Given this, I'm leaning towards adding a parameter "ready" in updateSnapshotStatusWithEvent to limit the scope of this change. This way we only change the behavior when snapshot class is not found.

I see that in the PV controller, it updates the phase from bound to available if volume.Spec.ClaimRef is nil.  There's a updateVolumePhaseWithEvent function which has phase and input parameter.

Reply:
Updated updateSnapshotStatusWithEvent to include a "ready *bool" input parameter.  If nil, ReadyToUse will not be changed.  
If snapshot class is not found, ReadyToUse will not be changed.
If updateSnapshotStatus failed, ReadyToUse will not be changed.
For all other cases, the code stays the same as before, meaning ReadyToUse will be False.